File: backend/src/models/DestinationBaseline.ts
import mongoose from 'mongoose';

const placeRefSchema = new mongoose.Schema({
  name: { type: String, required: true },
  placeId: { type: String },
  coordinates: {
    lat: { type: Number },
    lng: { type: Number },
  },
  category: { type: String },
  rating: { type: Number },
}, { _id: false });

const routeTemplateSchema = new mongoose.Schema({
  nights: { type: Number, required: true },
  cities: [{ type: String }],
  rationale: { type: String },
}, { _id: false });

const destinationBaselineSchema = new mongoose.Schema({
  key: { type: String, required: true, unique: true, index: true },
  name: { type: String, required: true },
  country: { type: String },
  placeId: { type: String },
  coordinates: {
    lat: { type: Number },
    lng: { type: Number },
  },
  popularityScore: { type: Number, default: 0 },
  topAttractions: { type: [placeRefSchema], default: [] },
  topRestaurants: { type: [placeRefSchema], default: [] },
  neighborhoods: { type: [placeRefSchema], default: [] },
  routeTemplates: { type: [routeTemplateSchema], default: [] },
  refreshedAt: { type: Date, default: Date.now },
  ttlDays: { type: Number, default: 30 },
}, { timestamps: true });

destinationBaselineSchema.index({ name: 'text', country: 'text' });

export const DestinationBaseline = mongoose.model('DestinationBaseline', destinationBaselineSchema);

File: backend-python/app/models/destination_baseline.py
"""DestinationBaseline ORM model."""

from datetime import datetime, timezone

from sqlalchemy import String, Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base, PortableJSON


class DestinationBaseline(Base):
    __tablename__ = "destination_baselines"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    key: Mapped[str] = mapped_column(String(200), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(200))
    country: Mapped[str | None] = mapped_column(String(100), nullable=True)
    place_id: Mapped[str | None] = mapped_column(String(200), nullable=True)
    coordinates: Mapped[dict | None] = mapped_column(PortableJSON, nullable=True)
    popularity_score: Mapped[int] = mapped_column(Integer, default=0)
    top_attractions: Mapped[list | None] = mapped_column(PortableJSON, nullable=True, default=list)
    top_restaurants: Mapped[list | None] = mapped_column(PortableJSON, nullable=True, default=list)
    neighborhoods: Mapped[list | None] = mapped_column(PortableJSON, nullable=True, default=list)
    route_templates: Mapped[list | None] = mapped_column(PortableJSON, nullable=True, default=list)
    refreshed_at: Mapped[datetime] = mapped_column(default=datetime.now(timezone.utc).replace(tzinfo=None))
    ttl_days: Mapped[int] = mapped_column(Integer, default=30)
    created_at: Mapped[datetime] = mapped_column(default=datetime.now(timezone.utc).replace(tzinfo=None))
    updated_at: Mapped[datetime] = mapped_column(default=datetime.now(timezone.utc).replace(tzinfo=None), onupdate=datetime.now(timezone.utc).replace(tzinfo=None))
